fairscale/nn/wrap/auto_wrap.py

```diff
@@ -34,26 +83,25 @@ def enable_wrap(module_blocklist: Optional[List] = None, **wrapper_kwargs: Any)
         with enable_wrap(**params):
             # Wraps layer in FSDP by default if within context
             self.l1 = wrap(torch.nn.Linear(5, 5))
-            # Wraps children modules by default based on min_num_params
-            self.l2 = auto_wrap(TransformerBlock(), min_num_params=1e8)
+            # Wraps children modules based on a different min_num_params
+            my_should_wrap = functools.partial(should_wrap, min_num_params=1e7)
+            self.l2 = auto_wrap(TransformerBlock(), shuold_wrap=my_should_wrap)
 
     Args:
-        module_blocklist: List of additional Module Classes to not wrap when
-            using :func:`auto_wrap`. This is useful to exclude unsupported
-            modules when wrapping recursively.
-        **wrapper_kwargs: Configuration settings that will be passed to all ``wrap``
+        should_wrap (Callable, Optional):
+            Custom function to control how to do :func:`auto_wrap`. This is
+            useful to exclude unsupported modules or wrap based on sizes when
+            wrapping recursively.
+            (default: :func:`should_wrap_default`)
+        **wrapper_kwargs:
+            Configuration settings that will be passed to all ``wrap``
             instances inside the context
     """
-    with ConfigAutoWrap(module_blocklist, **wrapper_kwargs):
+    with ConfigAutoWrap(should_wrap_default, **wrapper_kwargs):
         yield
 
 
-def wrap(
-    module: nn.Module,
-    cls: Callable = FullyShardedDataParallel,
-    activation_checkpoint: bool = False,
-    **wrap_overrides: Any
-) -> nn.Module:
+def wrap(module: nn.Module, activation_checkpoint: bool = False, **wrap_overrides: Any) -> nn.Module:
```

Comment:
similar to how you moved `cls` to `enable_wrap`, should we remove `activation_checkpoint`? In fairseq I found it more intuitive to do this in the context manager too.

For example:
```python
def my_wrapper(module, **wrap_overrides):
    return FSDP(activation_checkpoint(module), **wrap_overrides)

with enable_wrap(my_wrapper, ...):
    (...)
```

Btw, activation checkpointing works with FSDP as long as it's aligned to FSDP boundaries. This makes me wonder if we should add a `checkpoint_activations` convenience flag to FSDP. Then one could do:
```python
with enable_wrap(FSDP, checkpoint_activations=True):
    (...)
```

Reply:
I am fine either way. I slightly worry that we marry FSDP and checkpoint too early and removing some flexibility since people can do checkpointing inside a FSDP without wrapping. So, let me just first move checkpointing option into the context?

Reply:
Yep makes sense

Reply:
done. I like this a lot.